build system overhaul - use container images for testing

alanfranz-at-work opened the pull request:
build system overhaul - use container images for testing
    
While investigating an environment-dependent bug, I found it very hard
to reproduce the testing environment, since it was using github actions
with their own virtual machines content.
    
This approach should make debugging much faster on developer
environments: github actions uses a docker image for testing, and the
very same image can be built and used from the developer machine. The
github workflow only contains calls to scripts that can be executed in
the same way on a developer's machine.

This also fixes the issues that were detected in #77 and while trying to merge #79

Commit:
build system overhaul - use container images for testing

While investigating an environment-dependent bug, I found it very hard
to reproduce the testing environment, since it was using github actions
with their own virtual machines content.

This approach should make debugging much faster on developer
environments: github actions uses a container image for testing, and the
very same image can be built and used from the developer machine. The
github workflow only contains calls to scripts that can be executed in
the same way on a developer's machine.

The deps that we consider essential (mysql version, percona version,
python version) are properly pinned.

Commit:
fix: xtrabackup message parsing in xtrabackup 8.0.28

xtrabackup 8.0.28 slightly changed the output messages, which weren't
picked up properly by our system. 8.0.28 also slipped in automatically
as it was the latest version on upstream repo.

Commit:
fix failing test

The test would fail when invoked with a different interpreter. It was
changed to be more robust.

Commit:
fix: increase short timeouts

Some timeouts were too short and caused test flakiness.

Commit:
remove EOLed python 3.6

Commit:
cleanup accidentally left statements

Commit:
disable flaky test and add ticket BF-1198

As soon as container images were added, this test was found to be flaky.
It seems to depend on some condition which wasn't triggered in Github
Actions native container but gets triggered very often within the
container.

This deserves an additional investigation, but such failures would be
red herrings right now.

kmichel-aiven commented on .github/workflows/build.yaml:
```diff
@@ -82,68 +111,56 @@ jobs:
+      image: ghcr.io/aiven/myhoard-test:${{ github.sha }}
     strategy:
+      fail-fast: false
+      # this isn't a standard matrix because some version testing doesn't make sense
+      # e.g. you cannot use mysql .28 with percona .27, and there's no clean way of
+      # skipping matrix items that aren't meaningful
```
This should be possible: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixexclude

alanfranz-at-work replied:
Technically, it could be possible. But would it be better? Then you need to calculate what's being executed and what not. This approach just uses a "flat list" instead of a matrix, and I doubt we'd need to check all possible python versions (mysql vs percona versions seem fare more risky!)

kmichel-aiven commented on .github/workflows/build.yaml:
```diff
@@ -82,68 +111,56 @@ jobs:
+      image: ghcr.io/aiven/myhoard-test:${{ github.sha }}
     strategy:
+      fail-fast: false
+      # this isn't a standard matrix because some version testing doesn't make sense
+      # e.g. you cannot use mysql .28 with percona .27, and there's no clean way of
+      # skipping matrix items that aren't meaningful
```
This should be possible: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixexclude

kmichel-aiven commented on Makefile:
```diff
@@ -70,3 +70,33 @@ build-dep-fedora:
+
+.PHONY: dockertest-pytest
+dockertest-pytest:
+	docker run -it --rm -v "$(shell pwd):/src:ro" myhoard-test-temp /src/scripts/pytest-inside $(PYTEST_ARGS)
```
Missing newline at end of file.

alanfranz-at-work replied:
Do we have any standard about this? I'd feel silly to check and fix this manually. ```make lint``` doesn't catch this either.

kmichel-aiven replied:
pylint does it for python files (C0304), but we don't have a linter for non-python.
the standard is... [posix](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206) :D (and GitHub showing a red sign on the line)

Most text editors and IDEs have an option to fix that automatically instead of waiting for a linter:
 - Pycharm: Settings→Editor→General→Ensure every saved files ends with a line break (at the bottom)
 - VsCode: File→Preferences→Settings→Search for "insert final newline"
 - VIM: `:set fixendofline`

alanfranz-at-work replied:
Yeah, I know that it's posix, but I'm not sure we should focus on this kind of details in a review. This should be catched (and fixed!) by some automatic tool. I'll change those right now, btw.

kmichel-aiven commented on README.md:
```diff
@@ -745,6 +745,33 @@ Current phase of backup restoration. Possible options are these:
   automatically but it may fail repeatedly and analyzing logs to get more
   details regarding the failure is advisable.
 
+Running tests
+=============
```
Can you add a section about running the tests without docker too ?

alanfranz-at-work replied:
Makes sense, will do.

kmichel-aiven commented on build-image/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+ENV DEBIAN_FRONTEND="noninteractive"
+RUN apt-get update && apt-get -y install sudo lsb-release wget tzdata libsnappy-dev \
+    libpq5 libpq-dev software-properties-common build-essential rsync curl
+RUN add-apt-repository -y ppa:deadsnakes/ppa
```
Missing newline at end of file.

kmichel-aiven commented on scripts/create-user:
```diff
@@ -0,0 +1,7 @@
+adduser testme sudo
+echo "%sudo ALL=(ALL:ALL) NOPASSWD: ALL" > /etc/sudoers.d/sudonopasswd
+chmod 440 /etc/sudoers.d/sudonopasswd
+chown -R testme:testme .
```
Missing newline at end of file.

kmichel-aiven commented on test/test_controller.py:
```diff
@@ -285,7 +285,7 @@ def test_backup_state_from_removed_site_is_removed(default_backup_site, mysql_em
         assert not os.path.exists(file_name)
 
-
+@pytest.mark.skip(reason="This seems very flaky. See BF-1198")
```
I don't think we should have references to BF-* in this repo.

alanfranz-at-work replied:
uh, OK, that I didn't know, will double-check. Is this only for source code? Or for PRs/commit messages as well?

kmichel-aiven commented on test/test_util.py:
```diff
@@ -255,10 +257,19 @@ def test_encrypt_decrypt():
-        output = subprocess.check_output(["ps", "-x", "--cols", "1000", "-o", "pid,command"])
-        raise AssertionError(f"None of the commands included in ps output {output!r}")
-    assert myhoard_util.detect_running_process_id("certainlynosuchprocesscurrentlyrunning") is None
+class TestDetectRunningProcessId:
```
pytest does not recommend xunit-style test classes with setup/teardown functions, using fixtures is more native to the framework and composes better. (it's not possible to ask another fixture from the setup/teardown functions for instance).

alanfranz-at-work replied:
I suppose you're referring to this:

```While these setup/teardown methods are simple and familiar to those coming from a unittest or nose background, you may also consider using pytest’s more powerful [fixture mechanism](https://docs.pytest.org/en/6.2.x/fixture.html#fixture) which leverages the concept of dependency injection, allowing for a more modular and more scalable approach for managing test state, especially for larger projects and for functional testing. You can mix both fixture mechanisms in the same file but test methods of unittest.TestCase subclasses cannot receive fixture arguments.```

I'd not say it's not recommended, just that it's an alternative approach.

"more native to the framework"  -> doesn't mean better, though. What advantage would we get from this 
"composes better" -> we don't need composition here, it's a very small test, and I think I've seen plenty of examples where fixtures are abused and reused too many times, and their dependency chain can lead to an hard-to-identify test behaviour.

My take would be "use a fixture where a fixture is useful", not "use a fixture as the default approach". The test as written is very self-contained and quite easy to understand. I see no advantage in using a fixture with a cleanup action.

kmichel-aiven commented on .github/workflows/build.yaml:
```diff
@@ -1,7 +1,5 @@
 name: Build MyHoard
 
-permissions: read-all
-
```
This reverts the hardening PR: https://github.com/aiven/myhoard/pull/75
I doubt that's a good thing.

alanfranz-at-work replied:
This change was originally done to for the junit publish, for which I had got an explicit "go" from security. But I think this is unnecessary now, so, good catch.

Also, note that #75 says:
```Set the default workflow permissions to read - because this workflow does not need to modify anything in the project.```

So, it's a correct security-by-default measure, but it was "hardening", not "fixing a security problem". If we need to relax permissions for good reasons, I see nothing preventing this.

kmichel-aiven reviewed (dismissed):
A few small changes.

The PR should be renamed too, it's not a reproducible image: there are apt-get and wget inside the docker scripts that result in a different image every time it's built.

alanfranz-at-work commented:
> The PR should be renamed too, it's not a reproducible image: there are apt-get and wget inside the docker scripts that result in a different image every time it's built.

It's reproducible between the developer's environment and github actions. But I'm always very attentive to naming issues, so, I'll try to come up with a better naming and description.

Commit:
Allow running tests natively (without containers)

Add docs for running tests natively; also, make sure we don't modify the
host system's state unnecessarily at development time.

Commit:
improve output in case of failure

Commit:
fix pylint disable line

Commit:
Correct test and better output

If failing, the output of the test was actually wrong (toctou).

This test has a proper output.

Commit:
Clarify build steps

Commit:
fix whitespace
